core/parachain/pvf/workers.cpp

```diff
@@ -0,0 +1,238 @@
+                                              size_t) mutable {
+            WEAK_LOCK(self);
+            if (ec) {
+              return cb(ec);
+            }
+            auto len_res = scale::decode<uint32_t>(*len);
+            if (len_res.has_error()) {
+              return cb(len_res.error());
+            }
+            self->reading->resize(len_res.value());
+            boost::asio::async_read(
+                self->reader,
+                libp2p::asioBuffer(*self->reading),
+                [cb{std::move(cb)}, reading{self->reading}](
+                    boost::system::error_code ec, size_t) mutable {
+                  if (ec) {
+                    return cb(ec);
+                  }
+                  cb(std::move(*reading));
+                });
+          });
+    }
+  };
+
+  PvfWorkers::PvfWorkers(const application::AppConfiguration &app_config,
+                         common::MainThreadPool &main_thread_pool,
+                         std::shared_ptr<libp2p::basic::Scheduler> scheduler)
+      : io_context_{main_thread_pool.io_context()},
+        main_pool_handler_{main_thread_pool.handlerStarted()},
+        scheduler_{std::move(scheduler)},
+        exe_{exePath()},
+        max_{app_config.pvfMaxWorkers()},
+        timeout_{app_config.pvfSubprocessDeadline()},
+        worker_config_{
+            pvf_runtime_engine(app_config),
+            app_config.runtimeCacheDirPath(),
+            app_config.log(),
+            app_config.disableSecureMode(),
+        } {}
+
+  void PvfWorkers::execute(Job &&job) {
+    REINVOKE(*main_pool_handler_, execute, std::move(job));
+    if (free_.empty()) {
+      if (used_ >= max_) {
+        queue_.emplace(std::move(job));
+        return;
+      }
+      auto used = std::make_shared<Used>(*this);
+      auto process = std::make_shared<ProcessAndPipes>(*io_context_, exe_);
+      process->writeScale(
+          worker_config_,
+          [WEAK_SELF, job{std::move(job)}, used{std::move(used)}, process](
+              outcome::result<void> r) mutable {
+            WEAK_LOCK(self);
+            if (not r) {
+              return job.cb(r.error());
+            }
+            self->writeCode(
+                std::move(job), {std::move(process)}, std::move(used));
+          });
+      return;
+    }
+    findFree(std::move(job));
+  }
+
+  void PvfWorkers::findFree(Job &&job) {
+    auto it =
+        std::find_if(free_.begin(), free_.end(), [&](const Worker &worker) {
+          return worker.code_hash == job.code.code_hash;
+        });
+    if (it == free_.end()) {
+      it = free_.begin();
+    }
+    auto worker = std::move(*it);
+    free_.erase(it);
+    writeCode(std::move(job), std::move(worker), std::make_shared<Used>(*this));
+  }
+
+  PvfWorkers::Used::Used(PvfWorkers &self) : weak_self{self.weak_from_this()} {
+    ++self.used_;
+  }
+
+  PvfWorkers::Used::~Used() {
+    IF_WEAK_LOCK(self) {
+      --self->used_;
+    }
+  }
+
+  void PvfWorkers::writeCode(Job &&job,
```

Comment:
Did you measure how long this takes? Would placing code in a shared memory have a noticeable impact on pvf execution time?

Reply:
Didn't measure.
After #2140 child process will read compiled code from file.